src/Plugin/search_api/backend/SearchApiElasticsearchBackend.php

```diff
@@ -863,6 +870,33 @@ protected function parseSearchFacets(array $response, QueryInterface $query) {
     return $result;
   }
 
+  /**
+   * Allow custom changes before sending a search query to Elastic Search.
+   *
+   * This allows subclasses to apply custom changes before the query is sent to
+   * Solr.
+   *
+   * @param \Drupal\search_api\Query\QueryInterface $query
+   *   The \Drupal\search_api\Query\Query object representing the executed
+   *   search query.
+   */
+  protected function preQuery(QueryInterface $query) {
```

Comment:
@dakkusingh not really sure what is the reason to have empty methods here. 
With this implementation what we are going to achieve is to just call two additional methods that doesn't do anything. If we want to allow modules to alter the Elastic Search query we should invoke the events here that will do that. Can you please advice on this one?

Reply:
Hi Nikki, this will allow other modules to hook in and either alter the query or alter the results..

I mainly need it to extend the functionality of my "Search API Logger" module here https://www.drupal.org/project/search_api_logger  

see http://cgit.drupalcode.org/search_api_logger/tree/src/Plugin/search_api/backend/SearchApiLoggerElasticSearchBackend.php

This is replicating logic of 
http://cgit.drupalcode.org/search_api_solr/tree/src/Plugin/search_api/backend/SearchApiSolrBackend.php#n1022
http://cgit.drupalcode.org/search_api_solr/tree/src/Plugin/search_api/backend/SearchApiSolrBackend.php#n1053

Reply:
I agree, Events would be a better approach :) 
Since this how its done in search_api_solr, I have tried to keep it consistent..

```
...
      $this->moduleHandler->alter('search_api_solr_search_results', $results, $query, $result);
      $this->postQuery($results, $query, $result);
    }
    catch (\Exception $e) {
...
```

Reply:
@dakkusingh now I get your point. Yes, the implementation in search_api_solr seems really strange to me but I don't see a problem to merge this one right now and change it later on just to unblock you. Cheers mate!